app/src/ui/lib/os-dark-mode.ts

```diff
@@ -0,0 +1,34 @@
+import { remote } from 'electron'
+import {
+  getAutoSwitchPersistedTheme,
+  // ApplicationTheme,
+} from './application-theme'
+
+export function useOSDarkMode() {
+  subscribeToThemeChangedNotification()
+  return updateThemeBasedOnSystem()
+}
+
+function updateThemeBasedOnSystem() {
+  const useDarkMode = remote.systemPreferences.isDarkMode()
+  const automaticallySwitchTheme = getAutoSwitchPersistedTheme()
+
+  if (automaticallySwitchTheme) {
+    // const theme = useDarkMode ? ApplicationTheme.Dark : ApplicationTheme.Light
+    // Fire theme to dispatcher somehow :/
```

Comment:
@shiftkey I got stuck here.... I'm not quiet sure how to fire this back to the dispatcher

Reply:
I ended up going with a different approach to what you had in mind, and I'll walk through it to contrast - hopefully it makes sense!

First, I defined these two helpers around dark mode detection in `app/src/ui/lib/dark-theme.ts` - there were a few checks for this around, so centralizing it helped with my sanity:

```ts
import { remote } from 'electron'
import { isMojaveOrLater } from '../../lib/get-os'

export function supportsDarkMode() {
  if (!__DARWIN__) {
    return false
  }

  if (!isMojaveOrLater()) {
    return false
  }

  return true
}

export function isDarkModeEnabled() {
  if (!supportsDarkMode()) {
    return false
  }

  return remote.systemPreferences.isDarkMode()
}
```

Next, I created the `ThemeChangeMonitor` in `app/src/ui/lib/theme-change-monitor.ts ` whose sole job it is to report when the theme has changed. I think I ported all the logic over, but use this as a starting point for what you're looking to achieve:

```ts
import { remote } from 'electron'
import { ApplicationTheme } from './application-theme'
import { IDisposable, Disposable, Emitter } from 'event-kit'
import { supportsDarkMode, isDarkModeEnabled } from './dark-theme'

class ThemeChangeMonitor implements IDisposable {
  private readonly emitter = new Emitter()

  private subscriptionID: number | null = null

  constructor() {
    this.subscribe()
  }

  dispose() {
    this.unsubscribe()
  }

  private subscribe = () => {
    if (!supportsDarkMode()) {
      return
    }

    this.subscriptionID = remote.systemPreferences.subscribeNotification(
      'AppleInterfaceThemeChangedNotification',
      this.onThemeNotificationFromOS
    )
  }

  private onThemeNotificationFromOS = (event: string, userInfo: any) => {
    // if there is details in the payload from the OS about the theme change we should
    // use that rather than query again for the dark theme state
    const darkModeEnabled = isDarkModeEnabled()

    const theme = darkModeEnabled
      ? ApplicationTheme.Dark
      : ApplicationTheme.Light
    this.emitThemeChanged(theme)

    // do we need to re-subscribe to the event after it was raised? if not, these
    // lines are unnecessary and can be removed
    this.unsubscribe()
    this.subscribe()
  }

  private unsubscribe = () => {
    if (this.subscriptionID !== null) {
      remote.systemPreferences.unsubscribeNotification(this.subscriptionID)
    }
  }

  // subscribers call this function and pass in the callback they want to run
  public onThemeChanged(fn: (theme: ApplicationTheme) => void): Disposable {
    return this.emitter.on('theme-changed', fn)
  }

  private emitThemeChanged(theme: ApplicationTheme) {
    this.emitter.emit('theme-changed', theme)
  }
}

// this becomes our singleton that we can subscribe to from anywhere
export const themeChangeMonitor = new ThemeChangeMonitor()

// this ensures we cleanup any existing subscription on exit
remote.app.on('will-quit', () => {
  themeChangeMonitor.dispose()
})
```

Last, I have the ability to subscribe to this from anywhere in the app - I chose to do this in the `AppStore` constructor (because that's got easy access to the state) but you might have a better idea:

```ts
import { themeChangeMonitor } from '../../ui/lib/theme-change-monitor'

// ...

    themeChangeMonitor.onThemeChanged(theme => {
      if (this.automaticallySwitchTheme) {
        this.selectedTheme = theme
        this.emitUpdate()
      }
    })
```

Reply:
Makes perfect sense. Thanks for taking the time to explain the thought process. I was also thinking we would have to do something with an `Emitter` and a singleton. Just wanted to make sure I wasn't crazy/over thinking it. Ill work on the implementation soon.